tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_selectors.ts

```diff
@@ -602,8 +618,25 @@ export const getFocusedSourceFileContent = createSelector(
 );
 
 export const getFocusedSourceLineSpec = createSelector(
-  selectSourceCode,
-  (sourceCode: SourceCodeState): SourceLineSpec | null => {
-    return sourceCode.focusLineSpec;
+  selectDebuggerState,
+  (state: DebuggerState): SourceLineSpec | null => {
+    const stackFrames = getFocusedStackFramesHelper(state);
+    const focusedLineSpec = state.sourceCode.focusLineSpec;
+    if (
+      state.stickToBottommostFrameInFocusedFile &&
+      stackFrames !== null &&
+      focusedLineSpec !== null
+    ) {
+      return getBottommostStackFrameInFocusedFile(stackFrames, focusedLineSpec);
```

Comment:
This selector is a bit confusing now.  Before, if my source code had focusLineSpec (say for L15), the selector just returned L15.

Now, the selector says to return the line corresponding to the bottommost callframe in this file (say for L200), which is a very different contract.  Changing `sourceCode.focusLineSpec` used to always trigger this selector, but now it may not.

It might be easier to reason about if we keep the selector simple, and move logic related to updating the `sourceCode.focusLineSpec` into the reducer.  Wdyt?

Reply:
Good question. In order to implement the "stick to the bottommost frame when user navigates 
executions or graph ops", we need an extra level of indirection. This extra level can be implemented
in one of two ways: 
1. As extra logic in a selector, as is done here.
2. As extra logic in reducers.

The downside of 2 is that it needs to touch upon multiple reducers, including
- `executionDigestFocused`, to accommodate user navigation of executions
- `graphOpFocused`, to accommodate user navigation of executions
- `stackFramesLoaded`, because stack frames of executions and graph ops
   are loaded lazily. The bottommost-frame calculation often needs to wait for this
   lazy loading.

But 1 is logically simpler because it utilizes the selector's ability to automatically
listen to changes in relevant bits of the ngrx store.

So in balance, I think approach 1 (the current approach) is better. To address your 
concern about confusion, I made the following changes:
1. I renamed this selector to `getMaybeBottommostStickingSourceLineSpec`, so that
    it'll be clearer that this doesn't return the value of `state.sourceCode.focusLineSpec`
     always.
2. I added a doc string to `state.sourceCode.focusLineSpec` to clarify that it is what the
    user has focused on and the UI may choose to focus on a different frame depending
   on other states of the store.

Reply:
Hm, I see that the selector approach is convenient.  As mentioned, it involves less code, and if some new action X is introduced that can alter the current stack trace, things automatically work.

However, I'm still opposed to it (2), and would prefer the reducer (1) approach for these reasons:

Limits us in the future
- The "current stack frame seen by the user" is accessible by effects and views, but not reducers.  If we ever want to write a reducer that acts based on the current stack frame, it would be impossible(?) without duplicating code.
- For example, how might we build a feature that lets users press `ctrl+[` to move from the current stack frame to its parent?  The current stack frame !== `state.sourceCode.focusedLineSpec`, and we can't use `getMaybeBottommostStickingSourceLineSpec` from the reducer.  We'd have to duplicate the logic in this selector.

Makes the mental model complex
- It is misleading imo to have a `state.sourceCode.focusedLineSpec`, that does not represent the focused line seen by the user.  The meaning of this state property has now become `state.sourceCode.lastLineSpecFocusedFromStackTraceUI` or `state.sourceCode.lineSpecFocusedWhenIgnoringStickToBottom`, both of which are not easy to read or reason about.  The comment added in this PR helps, but does not fix the problem of a misleading name, imo.
- Does dispatching `actions.sourceLineFocused(sourceLineSpec1)` guarantee that `sourceLineSpec1` will become focused in the UI, or does it get overridden by stickToBottom logic in the selector?  I think that currently it is guaranteed to focus the provided line spec, but this 
- A selector named `getMaybeBottommostStickingSourceLineSpec` is hard to reason about.  If a view just wants the currently selected stack frame (as seen by the user), they first need to understand the sticking behavior, which they are not concerned about at all.

I'd prefer avoiding leaking the stickToBottom concept from reducers into selectors and views, if possible.  Duplicating a helper function across the 3 reducer fn's you mentioned seems maintainable, and has the benefit of making it clear to readers that `executionDigestFocused` may affect the current stack frame.

Might it be simple to touching the 3 reducers like so
```
on(executionDigestFocused, (state) => {
  // ...
  return {...state, sourceCode: {..., focusedLineSpec: getFocusedLineSpec(state)}};
}

on(stackFramesLoaded, (state) => {
  // ...
  return {...state, sourceCode: {..., focusedLineSpec: getFocusedLineSpec(state)}};
}

on(graphOpFocused, (state) => {
  // ...
  return {...state, sourceCode: {..., focusedLineSpec: getFocusedLineSpec(state)}};
}

function getFocusedLineSpec(state) {
  // does what getMaybeBottommostStickingSourceLineSpec is doing
}
```
?